src/core/net/netif.hpp

```diff
@@ -156,6 +157,47 @@ class NetifMulticastAddress : public otNetifMulticastAddress,
     bool Matches(const Address &aAddress) const { return GetAddress() == aAddress; }
 };
 
+class ExternalNetifMulticastAddress : public NetifMulticastAddress
+{
+    friend class Netif;
+    friend class LinkedList<ExternalNetifMulticastAddress>;
+
+public:
+    /**
+     * This method returns the next `ExternalNetifMulticastAddress` instance in the a linked list.
+     *
+     * @returns A pointer to the next `ExternalNetifMulticastAddress` instance.
+     *
+     */
+    ExternalNetifMulticastAddress *GetNext(void)
+    {
+        return static_cast<ExternalNetifMulticastAddress *>(const_cast<otNetifMulticastAddress *>(mNext));
```

Comment:
Do we also need the `GetNext()` here for the use by `Pool` class?
If so, can we see if we can define it as `private` here (since `LinkedList<Ext...>` is a friend already)?

Reply:
Yes, we need this for Pool to compile, otherwise we get this compile error:
```c++
In file included from openthread/src/core/net/netif.cpp:34:
In file included from openthread/src/core/net/netif.hpp:41:
openthread/src/core/common/linked_list.hpp:191:28: error: assigning to 'ot::Ip6::ExternalNetifMulticastAddress *' from incompatible type 'ot::Ip6::NetifMulticastAddress *'
            mHead = mHead->GetNext();
                    ~~~~~~~^~~~~~~~~
openthread/src/core/common/pool.hpp:105:45: note: in instantiation of member function 'ot::LinkedList<ot::Ip6::ExternalNetifMulticastAddress>::Pop' requested here
    Type *Allocate(void) { return mFreeList.Pop(); }
                                            ^
openthread/src/core/net/netif.cpp:335:38: note: in instantiation of member function 'ot::Pool<ot::Ip6::ExternalNetifMulticastAddress, 2>::Allocate' requested here
    entry = mExtMulticastAddressPool.Allocate();
                                     ^
```

Another fix might be changing the code of `Pop()` to use `static_cast`:
```c++
    Type *Pop(void)
    {
        Type *entry = mHead;

        if (mHead != nullptr)
        {
            mHead = static_cast<Type *>(mHead->GetNext());
        }

        return entry;
    }
```

Does it sounds logical to you? @abtink

Reply:
moved to private.

Reply:
I think, if defining `GetNext()` on `ExternalMulticastAddress` as `private` works out, it is a good solution. We let the `Pool`'s linked list to use this (being a `friend`) but not any one else.

I am not sure about changing `Pop()`. 

----

This actually exposes a limitation of the `Pool` class that its template `Type` needs to be linked-listable. I think there may be way to remove such a restriction (and make `Pool` even more flexible), but that's probably for a future PR/solution. :)